src/gtk/toga_gtk/window.py

```diff
@@ -87,8 +83,16 @@ def set_content(self, widget):
 
         self.native.add(self.layout)
 
+        # Make the window sensitive to size changes
         widget.native.connect('size-allocate', self.on_size_allocate)
 
+        # Set the widget's viewport to be based on the window's content.
+        widget.viewport = GtkViewport(widget.native)
+
+        # Add all children to the content widget.
+        for child in widget.interface.children:
+            child._impl.container = widget
```

Comment:
Here also, can we avoid using _impl in the implementation layer some how?

Reply:
This one is slightly convoluted. We're operating at the implementation layer; but the actual heirarchy is stored at the interface level. So, we're:

* operating on an implementation widget (the parent);
* asking for the list of toga-interface level children;
* getting the implementation for each child; and
* setting the container of the child to be the same as the parent.

It's unfortunate that we have to do the "go back up to the interface and back down to the implementation" dance - but there's no way to avoid this without reproducing the toga-interface heirarchy on the implementation layer (which would be redundant). 

If you want to put a spin on this: all we're doing is operating on the implementation layer from within the implementation layer - we just have to go via the interface to get there.

Reply:
I "agree" with @danyeaw that this is very confusing :-| .

>It's unfortunate that we have to do the "go back up to the interface and back down to the implementation" dance 
>(which would be redundant)

It may be redundant but it could clarify how everything works.

Reply:
I'm not sure it would. Reproducing the hierarchy at the implementation layer would introduce all sorts of possibilities for error - adding a widget to a different parent in the implementation and interface layer; removing a widget from one heirarchy but not another... for me, at least, it's confusing enough having to deal with 2 hierarchies as it is (interface and native) - adding a third, which would be a complete duplicate of interface layer anyway, just to avoid `obj.interface._impl` seems like asking for trouble.

I'll completely agree that more documentation of the internal structure is required, though.

Reply:
Ok, yes I understand.

I will think about this a bit more after lunch :-p, maybe adding more comments on that section would help ;-)

Reply:
I think you have convinced me after I tried to wrap my brain around this more today. Your description made the iteration through the children logic definitely more clear to me. 😄  Why is Box the only Gtk+ widget with a native object as a separate class (TogaBox)?

Reply:
Mostly because the GTK widgets are designed to be used with signals, rather than extension. In GTK, you can use a "stock" GTK button, and register a listener to a "clicked" signal. In Cocoa, you have to subclass button and add an `on_pressed` method on the subclass. 

GTKBox is the only widget where this isn't the case, because we have to get involved with the redraw process.